[expo-device] Add expo-device module

vivianzzhu91 opened the pull request:
# Why

Created a new module called `expo-device`, which serves as an API to get physical device information, such as phone carrier, free system storage, fingerprint setting. This would have a couple of use cases, such as when developers care about carrier, security, or system storage.

# How

We referenced the existing module written by the community. https://github.com/react-native-community/react-native-device-info

# Test Plan

We tested the modules on the Android and iOS development platforms with both real devices and simulators. To test that the packages were working we used the Sandbox expo app to test the values that our API should return.

ninjaachibi committed:
Create the highly anticipated Device module

Co-authored-by: Robert L <1998.robert@gmail.com>
Co-authored-by: Vivian Z <tongyu@usc.edu>

vivianzzhu91 committed:
remove viewManager

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceViewManager.java:
```diff
@@ -0,0 +1,41 @@
+import org.unimodules.core.ViewManager;
+import org.unimodules.core.interfaces.RegistryLifecycleListener;
+
+public class DeviceViewManager extends ViewManager<DeviceView> implements RegistryLifecycleListener {
```
Remove

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceView.java:
```diff
@@ -0,0 +1,12 @@
+package expo.modules.device;
+
+import android.content.Context;
+import android.view.View;
```
Remove this file

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DevicePackage.java:
```diff
@@ -0,0 +1,22 @@
+  @Override
+  public List<ViewManager> createViewManagers(Context context) {
+    return Collections.singletonList((ViewManager) new DeviceViewManager());
+  }
```
Delete this function

EvanBacon reviewed (changes requested):
Remove the view templates

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+      String stringValue = storage.toString();
+      promise.resolve(stringValue);
+    } catch (NullPointerException e) {
+      e.printStackTrace();
```
@ide do we need to print the stack trace here?

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+      promise.resolve(ipAddressString);
+    } catch (Exception e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
```
If we print stacks then we should find a unified way to do it

vivianzzhu91 replied:
Sure. I just removed the printStackTrace and decided to only log in Android console and promise rejection.

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+
+    String macAddress = getWifiInfo().getMacAddress();
+
+    String permission = "android.permission.INTERNET";
```
Does this overlap with the Permissions interface? We may need to add the following to the manifest
```xml
<uses-permission android:name="android.permission.INTERNET"></uses-permission>
```

vivianzzhu91 replied:
I already added this permission to the AndroidManifest.xml. This line is only checking if the app got the user's permission to access the Internet, so there's no overlapping I guess.

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+            macAddress = res1.toString();
+          }
+        }
+      } catch (Exception ex) {
```
Unhandled exception?

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+  public void isAirplaneModeAsync(Promise promise) {
+    boolean isAirPlaneMode;
+
+    isAirPlaneMode = Settings.Global.getInt(mContext.getContentResolver(), Settings.Global.AIRPLANE_MODE_ON, 0) != 0;
```
```suggestion
    boolean isAirPlaneMode = Settings.Global.getInt(mContext.getContentResolver(), Settings.Global.AIRPLANE_MODE_ON, 0) != 0;
```

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+
+  @ExpoMethod
+  public void hasSystemFeatureAsync(String feature, Promise promise) {
+    if (feature == null || feature == "") {
```
```suggestion
    if (feature.equals(null) || feature.equals("")) {
```

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+
+  @ExpoMethod
+  public void isPinOrFingerprintSetAsync(Promise promise) {
+    KeyguardManager keyguardManager = (KeyguardManager) mContext.getApplicationContext().getSystemService(Context.KEYGUARD_SERVICE); //api 16+
```
Do we need to guard this version?

vivianzzhu91 replied:
This KeyguardManager is to check whether this device is secured by a Pin/Password/Fingerprint. It didn't actually guard anything though.

EvanBacon commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+  }
+
+  @ExpoMethod
+  public void getUserAgentAsync(Promise promise) {
```
@ide we should make a note about deprecating this from `expo-constants`

vivianzzhu91 committed:
remove all viewtemplates in android

EvanBacon commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+                    resolver:(UMPromiseResolveBlock)resolve
+                    rejecter:(UMPromiseRejectBlock)reject) {
+  //some iOS privacy issues
+  NSString *address = @"02:00:00:00:00:00";
```
@ninjaachibi Maybe cite this

esamelson replied:
can we not get the MAC address reliably on iOS? if not, it's probably better just to not expose a method at all than to return a fake value

ide replied:
I agree with Eric -- better to maybe just return null and be very clear in the documentation.

EvanBacon replied:
The old approach to getting this value returns `02:00:00:00:00:00` (a cool security addition that was added) which is why this just returns the raw string.

EvanBacon commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+  LAContext *context = [[LAContext alloc] init];
+  BOOL isPinOrFingerprintSet = ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthentication error:nil]);
+#endif
+  resolve(@[[NSNumber numberWithBool:isPinOrFingerprintSet]]);
```
@sjchmiela can they just return `BOOL`?

sjchmiela replied:
BOOL, but wrapped inside NSNumber (so they’re right about that), but there is a shorthand for that – `@(isPinOrFingerPrintSet)` should do the work. Btw I’m not sure if ot would be possible to put BOOL into an array 🤔

EvanBacon commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+UM_EXPORT_METHOD_AS(isPinOrFingerprintSetAsync, isPinOrFingerprintSetAsyncWithResolver:(UMPromiseResolveBlock)resolve rejecter:(UMPromiseRejectBlock)reject)
+{
+#if TARGET_OS_TV
+  BOOL isPinOrFingerprintSet = false;
```
```suggestion
  BOOL isPinOrFingerprintSet = NO;
```

EvanBacon commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+  return totalSpace;
+}
+
+- (uint64_t) freeDiskStorage {
```
```suggestion
- (uint64_t)freeDiskStorage {
```

for all methods

EvanBacon commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,90 @@
+    ) !== -1
+  );
+}
+export async function getFreeDiskStorageAsync(): Promise<String> {
```
```suggestion
export async function getFreeDiskStorageAsync(): Promise<string> {
```

EvanBacon commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,90 @@
+    return await ExpoDevice.isAirplaneModeAsync();
+  } else {
+    return Promise.reject('This platform does not support this method');
+  }
```
```ts
if (!ExpoDevice.isAirplaneModeAsync) {
  throw new UnavailabilityError('expo-device', 'isAirplaneModeAsync')
} 
return await ExpoDevice.isAirplaneModeAsync();
```

EvanBacon commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,90 @@
+}
+
+export async function hasSystemFeatureAsync(feature: string): Promise<boolean | string> {
+  if (Platform.OS === 'android') {
```
Same as `isAirplaneModeAsync`

EvanBacon commented on packages/expo-device/src/Device.types.ts:
```diff
@@ -0,0 +1,289 @@
+export const devicesWithNotch = [
```
These are constants not types

ninjaachibi replied:
made a new file `DeviceConstants.ts` to contain constants

EvanBacon commented on packages/expo-device/src/ExpoDevice.ts:
```diff
@@ -0,0 +1,2 @@
+import { NativeModulesProxy } from '@unimodules/core';
+export default NativeModulesProxy.ExpoDevice;
```
Create `ExpoDevice.web.ts`

```tsx
export default {
  get name(): string {
    return "ExpoDevice";
  }
  // ...
}
```

vivianzzhu91 committed:
clean java code format and exception handler

ninjaachibi committed:
fix JS comments

ninjaachibi committed:
Merge branch '@expo-device' of github.com:expo/expo into @expo-device

ninjaachibi committed:
fix JS files

sjchmiela commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+           //           @"serialNumber": @"undefined4", // ANDROID ONLY
+           @"supportedABIs": @[[self getCPUType]],
+           @"systemName": currentDevice.systemName,
+           @"totalMemory": @(self.totalMemory),
```
I think it would be clearer if, since `totalMemory` is a method, you'd use `[self totalMemory]` notation.

sjchmiela commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+  
+  return @{
+           @"brand": @"Apple",
+           @"carrier": self.carrier ?: [NSNull null],
```
You could use `UMNullIfNil` here and there for consistency

sjchmiela commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+  return dispatch_get_main_queue();
+}
+
+- (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
```
Seems it's not needed, remove this and `ModuleRegistryConsumer` compliance

sjchmiela commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+#endif
+
+@interface EXDevice() {
+  WKWebView *webView;
```
Let's either use `{ properties }` or `@property`

sjchmiela commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+  dispatch_async(dispatch_get_main_queue(), ^{
+    __strong EXDevice *strongSelf = weakSelf;
+    if (strongSelf) {
+      if (!strongSelf.webViewUserAgent) {
```
I think you could coalesce these two ifs into one, `(nil).webViewUserAgent` should return `nil` without failing, also would make the promise resolve/reject even if `strongSelf` is `nil`

ninjaachibi replied:
@EvanBacon were there any edge cases caught by using two `if` statements?

sjchmiela commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+        [strongSelf->webView evaluateJavaScript:@"window.navigator.userAgent;" completionHandler:^(id _Nullable result, NSError * _Nullable error) {
+          if (error) {
+            reject(@"ERR_CONSTANTS", error.localizedDescription, error);
+            return;
```
not clearing `webview` even though task completes

sjchmiela commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+  int success = 0;
+  // retrieve the current interfaces - returns 0 on success
+  success = getifaddrs(&interfaces);
+  if (success == 0) {
```
if `success != 0` shouldn't we reject the promise?

ninjaachibi replied:
yep, push a new commit to account for this

sjchmiela commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+
+#define DeviceTypeValues [NSArray arrayWithObjects: @"Handset", @"Tablet", @"Tv", @"Unknown", nil]
+
+- (DeviceType)getDeviceType
```
Why don't we simplify handling device type and instead of defining, maintaining our own enum:
- create a method that would convert `UIUserInterfaceIdiom` to one of `Handset`, `Tablet`, `Tv`, `Unknown`
    `- (NSString *)deviceType`
- return its value in `constantsToExport`
- in `isTablet` method, which imho could be removed in favor of a simple `@"isTablet": @([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad),` compare interface idiom (not `DeviceType`)
- remove `DeviceTypeValues` macro and enum

ninjaachibi replied:
Thanks for the suggestions @sjchmiela! They are reflected in the updated commits. One question: 
> * in `isTablet` method, which imho could be removed in favor of a simple `@"isTablet": @([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad),` compare interface idiom (not `DeviceType`)
After I tried implement that change, `isTablet` returned `0` when I tested. This is probably because in Objective-C, the `==` operator implicitly returns `0` or `1` like in C, so I decided to put expression in the `- (bool) isTablet` method.

sjchmiela commented on packages/expo-device/package.json:
```diff
@@ -0,0 +1,33 @@
+
+,"_resolved": ""
+,"_integrity": ""
+,"_from": "expo-device@latest"
```
Please remove `_fields`, they're a leftover from `expo generate`

sjchmiela commented on packages/expo-device/package.json:
```diff
@@ -0,0 +1,33 @@
+  "author": "650 Industries, Inc.",
+  "license": "MIT",
+  "homepage": "https://github.com/expo/expo/tree/master/packages/expo-device",
+  "peerDependencies": {
```
I think right now the convention is to use 

https://github.com/expo/expo/blob/3ff22c99005ff476fc50b6dd91f7697a02234df2/packages/expo-constants/package.json#L34-L37

sjchmiela commented on packages/expo-device/package.json:
```diff
@@ -0,0 +1,33 @@
+  "description": "ExpoDevice standalone module",
+  "main": "build/Device.js",
+  "types": "build/Device.d.ts",
+  "scripts": {
```
moar scripts plz

https://github.com/expo/expo/blob/3ff22c99005ff476fc50b6dd91f7697a02234df2/packages/expo-constants/package.json#L8-L15

sjchmiela commented on packages/expo-device/package.json:
```diff
@@ -0,0 +1,33 @@
+  "peerDependencies": {
+    "@unimodules/core": "~1.0.0"
+  },
+  "gitHead": "f213c9be2a909dae42a0a27d8aa4b9eac8e97be1"
```
add devdependency on `expo-module-scripts`

https://github.com/expo/expo/blob/3ff22c99005ff476fc50b6dd91f7697a02234df2/packages/expo-constants/package.json#L41-L42

sjchmiela commented on packages/expo/package.json:
```diff
@@ -156,4 +156,4 @@
   },
   "gitHead": "78ebcdbae1d255a175eccb7183515110521607c1"
-}
+}
```
Do not remove newline at the end of file

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,87 @@
+
+export async function getUserAgentAsync(): Promise<string>{
+  return await ExpoDevice.getUserAgentAsync();
+}
```
add new line at the end of file (reconfigure your text editor)

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,87 @@
+  }
+  return await ExpoDevice.hasSystemFeatureAsync(feature);}
+
+export async function isPinOrFingerprintSetAsync(): Promise<boolean>{
```
Prettify this file with prettifier
```suggestion
export async function isPinOrFingerprintSetAsync(): Promise<boolean> {
```

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,87 @@
+  return await ExpoDevice.isAirplaneModeAsync();
+}
+
+export async function hasSystemFeatureAsync(feature: string): Promise<boolean | string> {
```
Is there a situation when this could resolve with a `string`?

ninjaachibi replied:
nope, changed it to only be a `boolean`

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,87 @@
+export const supportedABIs = ExpoDevice.supportedABIs;
+export function hasNotch(): boolean {
+  return (
+    devicesWithNotch.findIndex(
```
`Array.some`! https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

sjchmiela commented on packages/expo-device/src/DeviceConstants.ts:
```diff
@@ -0,0 +1,289 @@
+    },
+];
+
+export const deviceNamesByCode = {
```
Maybe split this file into two so it's easier to read/search?

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,87 @@
+
+import { Platform, UnavailabilityError } from '@unimodules/core';
+
+export const brand = ExpoDevice.brand;
```
Writing exports this way if, by any chance, this code would end up being run in an application that wouldn't have `ExpoDevice` available, the application would crash immediately. Add `ExpoDevice ? ExpoDevice.brand : null`, if you want to export these constants from this file

ninjaachibi replied:
awesome, thanks

sjchmiela commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,230 @@
+## API
+
+```js
+import { Device } from 'expo';
````
Should we advertise this way of importing in new documentation?

vivianzzhu91 replied:
Would 'import * as Device from 'expo-device';' better than this?

ninjaachibi replied:
what alternative would you suggest?

brentvatne replied:
indeed, `import * as Device from 'expo-device'` is preferred. in sdk34 we will remove imports of this style from the expo package

ninjaachibi replied:
What would the alternative be?

sjchmiela commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,230 @@
+
+### Constants
+
+#### `Device.brand: string`
```
Would it be possible to make this list more concise (as a list for example?), maybe with some examples? So that users know what's the difference between `brand`, `manafacturer` and `model`

sjchmiela commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,230 @@
+
+#### `Device.model: string`
+
+**iOS warning**: The list with device names is maintained by the community and could lag new devices. It is recommended to use `getDeviceId()` since it's more reliable and always up-to-date with new iOS devices. We do accept pull requests that add new iOS devices to the list with device names.
```
There's no `getDeviceId` 😅

sjchmiela commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,230 @@
+**Examples**
+
+```js
+const hasNotch = Device.hasNotch();
````
```suggestion
import { hasNotch } from 'expo-device';

hasNotch(); // true or false
```

sjchmiela commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,230 @@
+});
+```
+
+### `Device.getIPAddressAsync()`
````
Judging by eg. [`Linking.makeUrl`](https://docs.expo.io/versions/v33.0.0/sdk/linking/#linkingmakeurlpath-queryparams) method, I think that abbreviations should end up camelcased, so shouldn't it be `getIpAddressAsync`?

sjchmiela reviewed (changes requested):
Congratulations on your contribution! I can't wait for the users to be able to use this feature. I reviewed your PR and left some comments, could you go through them and apply suggestions if you think they make sense?

vivianzzhu91 commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+  public DeviceModule(Context context) {
+    super(context);
+    mContext = context;
+    deviceType = this.getDeviceType(context);
```
I think we have a constant called "deviceType" that requires this value in the getBundledConstants().

sjchmiela replied:
We could fetch that value then, couldn't we? 🤔 (If you're concerned about the fact that constructor will be called only once per an object and `getConstants` could be called multiple times, don't worry! You can depend on `getConstants` being called only once per object.)

vivianzzhu91 commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+    mModuleRegistry = moduleRegistry;
+    mActivityProvider = moduleRegistry.getModule(ActivityProvider.class);
+    mActivity = mActivityProvider.getCurrentActivity();
+    mPhoneNumber = getPhoneNumber();
```
This is a "phoneNumber" constant in the module and it requires a read_phone_state permission. For such dangerous permission, it would ask for such permission at runtime. So in order to get the permission before the user tries to get the phone number, I need to ask for the permission when initialized.

sjchmiela replied:
If I understand you correctly, calling `getPhoneNumber()` triggers permissions dialog, right? I have doubts whether asking for the permission once the application starts is the best way of doing this—often apps want to first explain to users why they need certain personal data and only then ask/not ask for it.

So because
- it requires a runtime permission dialog and we want to show it only when the user asks
- theoretically phone number could change during runtime,

1. `phoneNumber` can't be in constants,
2. `phoneNumber` can't be a cached variable,
3. we may want to expose `getPhoneNumberAsync` and explicitly state in the documentation it will trigger permissions dialog

vivianzzhu91 replied:
Doing a getPhoneNumberAsync() is indeed more logical.

vivianzzhu91 replied:
I took a closer look at the logic, but I think it would be like this (reference from https://developer.android.com/training/permissions/requesting): 

1. The user calls `getPhoneNumberAsync()`
2. Check for permission of `READ_PHONE_STATE`
3. If not granted, request the permission
1) granted after request, get `mPhoneNumber` in current state.
2) not granted, `mPhoneNumber` = informative message(permission is not granted)
4.If already granted, `mPhoneNumber` would fetch the phoneNumber value again.

So in this way, keeping a private variable `mPhoneNumber` seems reasonable since we need to store the phoneNumber value after permission request grant the permission and need to access it through `getPhoneNumberAsync()`. And the `mPhoneNumber` value would update each time since we fetch the value when the user calls the `getPhoneNumberAsync()`.

I'm not sure if this makes any sense or not.

vivianzzhu91 commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+  public Map<String, Object> getConstants() {
+    HashMap<String, Object> constants = new HashMap<>();
+
+    Bundle bundle = getBundledConstants();
```
This is a transformation from the bundle constants to the hashMap objects of the existing constants so that the js layer could access.

sjchmiela replied:
Why don't we create a HashMap in the first place?

vivianzzhu91 commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+    Bundle constants = new Bundle();
+
+    constants.putString("brand", Build.BRAND);
+    constants.putString("carrier", this.getCarrier());
```
Sure! We should do a getCarrierAsync() function instead of exporting it as a constant.

ninjaachibi committed:
Update docs/pages/versions/unversioned/sdk/device.md

Co-Authored-By: Stanisław Chmiela <sjchmiela@users.noreply.github.com>

vivianzzhu91 committed:
change isAirplaneModeEnabledAsync() name and carrier constants

ninjaachibi committed:
remove unnecessary lines

ninjaachibi committed:
fix package.json

ninjaachibi committed:
Update packages/expo-device/ios/EXDevice/EXDevice.m

Co-Authored-By: Stanisław Chmiela <sjchmiela@users.noreply.github.com>

vivianzzhu91 commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+  private WifiInfo getWifiInfo() {
+    if (this.wifiInfo == null) {
+      WifiManager manager = (WifiManager) mContext.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
+      this.wifiInfo = manager.getConnectionInfo();
```
I'd choose to obtain the WifiInfo each time requesting the IP address so that we do not cache this value. Thanks for catching it!

ninjaachibi committed:
fix exports and newline

ninjaachibi committed:
add array.some for devicesWithNotch()

vivianzzhu91 committed:
update android function name and handle exceptiont

ninjaachibi committed:
fix package.json

ninjaachibi committed:
add ternary

ninjaachibi committed:
Merge branch 'master' into @expo-device

vivianzzhu91 committed:
update device docs with changed function and constant

vivianzzhu91 committed:
change DeviceConstants into two files and add ExpoDevice.web.ts

ninjaachibi commented on packages/expo-device/ios/EXDevice/DeviceUID.h:
```diff
@@ -0,0 +1,18 @@
+//  Created by Robert Luo on 6/18/19.
+//
+
+#ifndef DeviceUID_h
```
nope don't think it does anything.

vivianzzhu91 commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,385 @@
+    mPhoneNumber = this.getPhoneNumber();
+    constants.putString("phoneNumber", mPhoneNumber);
+    constants.putString("serialNumber", this.getSerial());
+    constants.putString("systemName", "Android");
```
In the react native device info reference, they hard-coded this systemName for all android functions. And there's no available functions to get system name, so maybe this is hardcoded?

sjchmiela replied:
I haven't checked, but do you know what https://developer.android.com/reference/android/os/Build.VERSION.html#BASE_OS returns? I can see it raised some issues in the past (https://stackoverflow.com/questions/53541961/build-version-base-os-returns-empty-string-on-android-8), maybe `Android` could be a fallback? 🤔

vivianzzhu91 replied:
I could check the version first, and then if returns an empty string from the BASE_OS, then I'll just add "Android".

ninjaachibi committed:
[ios] refactoring names

ninjaachibi committed:
[ios] move getCarrier to async method

ninjaachibi committed:
change IPAddress -> [camelcase] IpAddress

ninjaachibi committed:
[ios] fix strong typing, make properties private

ninjaachibi committed:
IPAddress -> [camelcase] IpAddress

ninjaachibi committed:
[ios] use method notation

vivianzzhu91 committed:
move phoneNumber from constant to getPhoneNumberAsync()

vivianzzhu91 committed:
update getPhoneNumberAsync() in docs

vivianzzhu91 committed:
update importing Device in docs

vivianzzhu91 committed:
rebuild expo-device package and update build/Device.js.map

sjchmiela commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,363 @@
+    constants.put("manufacturer", Build.MANUFACTURER);
+    constants.put("model", Build.MODEL);
+    constants.put("serialNumber", this.getSerial());
+    String systeName = "";
```
Extract this to a private method please and `systeName` ➡️ `systemName`

sjchmiela commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,363 @@
+    actMgr.getMemoryInfo(memInfo);
+    constants.put("totalMemory", memInfo.totalMem);
+
+    mDeviceType = getDeviceType(mContext);
```
Make it a local variable, not instance variable

sjchmiela commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,363 @@
+        int permissionCheck = ContextCompat.checkSelfPermission(mContext, Manifest.permission.READ_PHONE_STATE);
+
+        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
+          ActivityCompat.requestPermissions(mActivity, new String[]{Manifest.permission.READ_PHONE_STATE}, REQUEST_READ_PHONE_STATE);
```
If `getSerial()` triggers a permission dialog it cannot be called in `getConstants` — we don't want to show permission dialogs before app code asks us to!

sjchmiela commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,363 @@
+        promise.reject(e);
+      }
+    }
+    promise.resolve(macAddress);
```
If some permission is not granted we resolve with an empty string? Shouldn't we reject here? 🤔

sjchmiela commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,363 @@
+  @ExpoMethod
+  public void getCarrierAsync(Promise promise) {
+    try {
+      promise.resolve(this.getCarrier());
```
I would move `getCarrier` implementation here (in this case I would say that extracting to a separate method is unnecessary), but that's just a suggestion

sjchmiela replied:
btw `getCarrier` at the moment already handles any exceptions, so the `catch` below is unnecessary

sjchmiela commented on packages/expo-device/ios/EXDevice/DeviceUID.m:
```diff
@@ -0,0 +1,148 @@
+//
```
Remove this and implementation file

ide replied:
This whole class seems unnecessary -- just use `identifierForVendor` which we can reliably call.

sjchmiela commented on packages/expo-device/src/DeviceNameByCode.ts:
```diff
@@ -0,0 +1,90 @@
+export const deviceNamesByCode = {
+    'iPod1,1': 'iPod Touch', // (Original)
```
Why 4-space indentation? Do you have prettier set up properly?

sjchmiela commented on packages/expo-device/src/ExpoDevice.web.ts:
```diff
@@ -0,0 +1,5 @@
+  get name(): string {
+    return "ExpoDevice";
+  }
+}
```
I think atm the convention is to add new lines at the end of files — please ensure you've prettified the files

sjchmiela reviewed (commented):
Rereviewed! 🙂

ninjaachibi committed:
[ios] Add UMNullIfNil notation, fix method spacing

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,279 @@
+
+  Gets the device ID.
+
+- `Device.totalMemory: number`
```
It's a stupid error I made during the test. Sorry about that!

vivianzzhu91 committed:
update docs format and consistency

ninjaachibi committed:
[ios] remove moduleregistry compliance

ninjaachibi committed:
[ios] fix getUserAgentAsync

[ios] Add webView as a property

ninjaachibi committed:
[ios] rename getStorageDictionary -> documentFileSystemAttributes

vivianzzhu91 commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,363 @@
+            && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+            try{
+              TelephonyManager telMgr = (TelephonyManager) mContext.getApplicationContext().getSystemService(Context.TELEPHONY_SERVICE);
+              mPhoneNumber = telMgr.getLine1Number();
```
In order to communicate with the promise in `getPhoneNumberAsync` in `onRequestPermissionsResult, I would throw an `AccessControlException` so that in `getPhoneNumberAsync` I would catch it and reject the promise if failed to acquire the permission.

I tried to use `permissionsModule.askForPermissions` in `ImagePickerModule`, but got into `InvocationTargetException` every time I tried to acquire permission. Couldn't figure that out, so I used throw exception to reject the promise.

esamelson replied:
discussed offline, commenting here for reference 🙂  - need to add this permission to PermissionsModule

vivianzzhu91 committed:
format the ts files, serialNumber to getSerialNumberAsync

vivianzzhu91 committed:
change request permission logic, add getSerialAsync function

ninjaachibi committed:
[ios] Add promise rejection to IpAddressAsync

If no network interface can be retrieved, we should reject.

vivianzzhu91 committed:
update getSerialAsync in docs

ninjaachibi committed:
[ios] fix storage

[ios] Add elvis to totalMemory

[ios] convert storage to NSNumbers

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+  return [[NSFileManager defaultManager] attributesOfFileSystemForPath:[paths lastObject] error: nil];
+}
+
+- (uint64_t) totalDiskCapacity {
```
- I updated this to return `NSNumber *` where appropriate.

>  I think (calling methods on nil is safe, idk what about accessing properties like this, you'd need to check)
- I tried to hardcode `- (NSDictionary *)documentFileSystemAttributes ` to return `nil` to test this theory, but at runtime, it made all the other `constantsToExport` return `undefined`. I'm not sure if I'm missing something here, but it seems like there may be an issue with accessing properties on `nil`. So to be safe I decided to just first check that `storage` is not `nil`, then return the `FileSystemSize`, otherwise return `nil` for the function.

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+
+- (NSDictionary *) getStorageDictionary {
+  NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
+  return [[NSFileManager defaultManager] attributesOfFileSystemForPath:[paths lastObject] error: nil];
```
In the devices I've tested with, I've only had one object in the `paths`, is there a potential case in which the array could be greater than length 1?

vivianzzhu91 committed:
add READ_PHONE_STATE in permissionModule and update permission logic in DeviceModule

ninjaachibi committed:
[ios] fix DeviceType

* create a method that would convert `UIUserInterfaceIdiom` to one of `Handset`, `Tablet`, `Tv`, `Unknown`
    `- (NSString *)deviceType`

* remove `DeviceTypeValues` macro and enum

* change implementation of `isTablet` method

ninjaachibi committed:
Update packages/expo-device/ios/EXDevice/EXDevice.m

format documentFileSystemAttributes

Co-Authored-By: Stanisław Chmiela <sjchmiela@users.noreply.github.com>

ninjaachibi committed:
[ios] move WKWebView *webView into getUserAgentAsync()

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,252 @@
+    if (strongSelf) {
+      if (!strongSelf.webViewUserAgent) {
+        // We need to retain the webview because it runs an async task.
+        strongSelf->webView = [[WKWebView alloc] init];
```
I think that works, I've pushed a commit to reflect this. Let me know what you think.

vivianzzhu91 committed:
[android/js] remove getPhoneNumberAsync due to permissionListen and security concern

vivianzzhu91 committed:
[docs] remove getPhoneNumberAsync

esamelson commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,332 @@
+
+    int permissionCheck = ContextCompat.checkSelfPermission(mContext, Manifest.permission.READ_PHONE_STATE);
+    if (permissionCheck == PackageManager.PERMISSION_GRANTED) {
+      try {
```
extract this part to a separate method so you don't have to repeat it below

esamelson commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,332 @@
+        }
+      } catch (SecurityException e) {
+        Log.e(TAG, e.getMessage());
+        promise.reject(e);
```
add error codes to all of these promise rejections (`ERR_DEVICE_(something about the method)`)

esamelson commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,332 @@
+              promise.reject(e);
+            }
+          } else {
+            promise.reject(new SecurityException("User rejected permissions"));
```
no need to create a SecurityException - you can just reject with an error code and a message.

esamelson commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,332 @@
+      long blockSize;
+
+      availableBlocks = external.getAvailableBlocksLong();
+      blockSize = external.getBlockSizeLong();
```
you can define these variables at the same time you declare them, on line 215-216

esamelson commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,332 @@
+
+  @ExpoMethod
+  public void getMACAddressAsync(String interfaceName, Promise promise) {
+    String permission = "android.permission.INTERNET";
```
this should be a constant defined by the Android OS already

vivianzzhu91 replied:
The constant value is this string `"android.permission.INTERNET"` ( reference from https://developer.android.com/reference/android/Manifest.permission)

esamelson commented on packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java:
```diff
@@ -0,0 +1,332 @@
+        promise.reject(e);
+      }
+    } else {
+      promise.reject(new AccessControlException("No permission granted to access the Internet"));
```
same as above - no need to create a new AccessControlException, instead just pass in an error code and message directly to promise.reject.

esamelson commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,256 @@
+    return;
+  }
+  
+  else if (success == 0) {
```
fmt -- put this back up on line 111. also, is the if clause here needed?

esamelson commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,256 @@
+
+- (bool) isTablet
+{
+//  return [[self deviceType] isEqualToString:@"Tablet"];
```
is this still needed?

esamelson commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,256 @@
+- (NSDictionary *)constantsToExport
+{
+  UIDevice *currentDevice = [UIDevice currentDevice];
+  NSString *uniqueId = [EXDeviceUID uid]; //TODO: need to import this
```
is this todo still relevant? (same question about all of the comments in the dictionary being returned)

esamelson commented on packages/expo-device/ios/EXDevice/EXDeviceUID.m:
```diff
@@ -0,0 +1,144 @@
+
+#pragma mark - UID Generation methods
+
++ (NSString *)appleIFV {
```
I'm fairly sure Apple restricts the use of this ID pretty heavily, and I'm not sure we want to include it here because it can cause apps (even apps that aren't using it in JS) to be flagged during Apple review.

ide replied:
I think this ID is OK actually -- it's not the IDFA (nor UDID but that is no longer exposed except through the provisioning profile dance). But maybe it does make sense to compartmentalize the fields we think are riskier (I'm suspect of the the CoreTelephony ones, though haven't done any research to see if they cause issues).

esamelson reviewed (changes requested):
this is looking good! added a few more comments. one other thing is that it looks like some of this code might be copied directly from the react-native-device-info package. that's fine but we should add a copy of their license to the end of ours.

vivianzzhu91 commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,101 @@
+}
+
+export async function getMACAddressAsync(interfaceName: string): Promise<string> {
+  if (Platform.OS === 'ios') {
```
Yes, the function would be android specific that the user can choose which interface to get the MacAddress. Great logic

vivianzzhu91 committed:
[ts/js] modify variable declaration and function logic

vivianzzhu91 committed:
[android] update promise rejection error codes and declare private helper

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+#### Arguments (Android Only)
+
+- **interfaceName (_string_)** -- A string representing interface name (`eth0`, `wlan0`) or `null`, meaning the method should fetch MAC address of the first available interface. (On iOS this argument is ignored.)
```
What about `undefined`?

vivianzzhu91 replied:
I just modified the changes to the java code to reject the promise with `undefined interface name` if such interfaceName is not available on `NetworkInterface.getNetworkInterfaces()`.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+#### Arguments (Android Only)
+
+- **interfaceName (_string_)** -- A string representing interface name (`eth0`, `wlan0`) or `null`, meaning the method should fetch MAC address of the first available interface. (On iOS this argument is ignored.)
```
```suggestion
- **interfaceName (_string_)** -- A string representing interface name (`eth0`, `wlan0`) or `null`, meaning the method should fetch the MAC address of the first available interface. (On iOS this argument is ignored.)
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+#### Returns
+
+A Promise that resolves to a string of the network adapter MAC address or empty string if there's no such address matching the interface.
```
- This should return null instead of an empty string in the edge case scenarios.
- What if there is no interface with the given name?

vivianzzhu91 replied:
I'll return null if the mac address is null with the given interface name.
If there's no matching mac address with such interface name passed in, we'll reject the promise with message `Undefined interface name`.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+Gets the device serial number.
+
+Would prompt the user for permission to read the phone to access the phone number.
```
A "phone number" sounds like it returns `(xxx) xxx-xxxx` -- reword this and be more specific and clear.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,273 @@
+**Examples**
+
+```js
+Device.isAirplaneModeEnabledAsync().then(airPlaneModeOn => {
````
```suggestion
Device.isAirplaneModeEnabledAsync().then(airplaneModeOn => {
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+### `Device.hasSystemFeatureAsync(feature)` (Android Only)
+
+Tells if the device has a specific system feature.
```
What does this mean? It's not clear what "system feature" means and how you would figure out what argument values to pass in.

vivianzzhu91 replied:
I added another function called `getSystemAvailableFeaturesAsync` to return all available system features on the device as a string[]. Will add it to the docs as well.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,273 @@
+await Device.hasSystemFeatureAsync('amazon.hardware.fire_tv'); // true or false
+```
+
+### `Device.isPinOrFingerprintSet()`
````
This method name seems short-sighted e.g. what if the device has other types of biometrics like Face ID for the lock screen?

vivianzzhu91 replied:
Renamed to `hasLocalAuthenticationAsync`.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+#### Returns
+
+Returns a Promise that resolves the `boolean` value for whether the device has set a Pin or Fingerprint.
```
```suggestion
Returns a Promise that resolves the `boolean` value for whether the device has set a PIN or fingerprint to unlock the device.
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+### `Device.isPinOrFingerprintSet()`
+
+Tells if a PIN number or a fingerprint was set for the device.
```
```suggestion
Tells if a PIN or a fingerprint was set for the device.
```

ide replied:
In what contexts is the PIN / fingerprint / etc used? The lock screen? Payments? Be more comprehensive.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,273 @@
+  Device.supportedABIs; // [ "arm64 v8", "Intel x86-64h Haswell", "arm64-v8a", "armeabi-v7a", "armeabi" ]
+  ```
+
+### Methods
````
This section seems redundant with the actual method documentation below -- remove?

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+  Gets the device brand.
+
+  - iOS: `"Apple"`
+  - Android: eg. `"Xiaomi"`
```
```suggestion
  - Android: e.g., `"Xiaomi"`
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+## Installation
+
+This API is pre-installed in [managed](../../introduction/managed-vs-bare/#managed-workflow) apps. It is not yet available for [bare](../../introduction/managed-vs-bare/#bare-workflow) React Native apps.
```
(What do we need to do to make it available for the bare workflow?)

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+title: Device
+---
+
+Provide information of devices on the application.
```
Be more detailed: what kind of information? When would you want to use this module and continue reading the rest of this document?

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+- `Device.model: string`
+
+  **iOS warning**: The list with device names is maintained by the community and could lag new devices. It is recommended to use `deviceId` since it's more reliable and always up-to-date with new iOS devices. We do accept pull requests that add new iOS devices to the list with device names.
```
This looks copied straight from react-native-device-info -- write original text.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+  Gets the device model.
+
+  - iOS: eg. `"iPhone XS Max"`
```
```suggestion
  - iOS: e.g., `"iPhone XS Max"`
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+  Gets the device model.
+
+  - iOS: eg. `"iPhone XS Max"`
+  - Android: eg. `"Pixel 2"`
```
```suggestion
  - Android: e.g., `"Pixel 2"`
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+- `Device.deviceId: string`
+
+  Gets the device ID.
```
What is the "device ID"?

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+- `Device.systemName: string`
+
+  Gets the device OS name.
```
Provide examples

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+- `Device.totalMemory: number`
+
+  Gets the device total memory, in bytes.
```
```suggestion
  Gets the device's total memory, in bytes.
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+- `Device.carrier: string`
+
+  Gets the carrier name (network operator).
```
```suggestion
  Gets the carrier name (network operator).
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+- `Device.supportedABIs: string[]`
+
+  Returns a list of supported processor architecture version
```
```suggestion
  Returns a list of supported processor architecture versions.
```

ide commented on android/expoview/src/main/java/host/exp/exponent/utils/PermissionsHelper.java:
```diff
@@ -170,6 +171,8 @@ private int permissionToResId(String permission) {
         return R.string.perm_fine_location;
       case android.Manifest.permission.ACCESS_COARSE_LOCATION:
         return R.string.perm_coarse_location;
+      case android.Manifest.permission.READ_PHONE_STATE:
```
READ_PHONE_STATE is marked as "dangerous" in the Android documentation; we probably should not add this to all apps that use expo-device: https://developer.android.com/reference/android/Manifest.permission.html#READ_PHONE_STATE

vivianzzhu91 replied:
So shall we remove the `getSerialNumberAsync` then? Because that method requires `READ_PHONE_STATE` permission in runtime, so there might be no way to access the serial number without this permission.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+### `Device.getIpAddressAsync()`
+
+Gets the device current IP address.
```
Also seems like it'd be more apropos to put the IP- & MAC-address methods into a networking module. Perhaps the airplane-mode check belongs there too. (@brentvatne what do you think about creating a derivative of r-n-netinfo like `expo-network-info` so that (a) we don't need to keep up to date with netinfo and also can put the IP/MAC-address methods in there?)

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+#### Returns
+
+A boolean that represents the support for notch display.
```
```suggestion
A boolean value that specifies whether the device's main display has a notch.
```

ide commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,256 @@
+#import <EXDevice/DeviceUID.h>
+#import <UMCore/UMUtilities.h>
+
+#include <ifaddrs.h>
```
`#import`

ide commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,256 @@
+
+- (NSString *)carrier
+{
+#if (TARGET_OS_TV)
```
Remove for now (we'll need to add something similar back when we support Catalyst for Mac)

ide commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,256 @@
+           @"deviceType": UMNullIfNil([self deviceType]),
+           @"deviceName": currentDevice.name, //TODO, ADD TO JS AS WELL
+           @"deviceId": UMNullIfNil([self deviceId]),
+           //           @"freeDiskStorage": @(self.freeDiskStorage),
```
Delete these commented-out fields

ide commented on packages/expo-device/ios/EXDevice/EXDeviceUID.m:
```diff
@@ -0,0 +1,144 @@
+
++ (NSString *)appleIFV {
+  if(NSClassFromString(@"UIDevice") && [UIDevice instancesRespondToSelector:@selector(identifierForVendor)]) {
+    // only available in iOS >= 6.0
```
We support only iOS 10+

ide commented on packages/expo-device/src/DeviceWithNotchConst.ts:
```diff
@@ -0,0 +1,226 @@
+export const devicesWithNotch = [
```
@brentvatne will there be a better way to know if the device has a notch with LayoutContext (or whatever Janic's work is called)? Seems important too if a device has multiple screens and we want to know if the current screen has a notch.

ide commented on packages/expo-device/package.json:
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "expo-device",
+  "version": "1.0.0",
```
```suggestion
  "version": "0.0.1",
```

Start with a version number that communicates this is a preview until it's ready for 1.0.

ide commented on packages/expo-device/package.json:
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "expo-device",
+  "version": "1.0.0",
+  "description": "ExpoDevice standalone module",
```
Write something more descriptive.

ninjaachibi committed:
[ios] Remove DeviceUID files

ninjaachibi committed:
Merge branch 'master' into @expo-device

ninjaachibi committed:
[ios] Fix EXDeviceUID import

- fix import name
- remove comments that were used for notes

ninjaachibi committed:
[ios] commit generated pods

ninjaachibi committed:
[ios] Change pbxproj

vivianzzhu91 committed:
[android] add getAvailablefeature and remove dangerous permission request function

vivianzzhu91 committed:
[docs] clarify definition and examples, remove and add functions

vivianzzhu91 committed:
[ts/js] update removal of getSerialNumber and add getSystemAvailableFeaturesAsync

vivianzzhu91 committed:
[js] update package.json with description and preview version

ninjaachibi committed:
[ios] remove comments

vivianzzhu91 committed:
[android] revert READ_PHONE_STATE permission in PermissionModule

ninjaachibi commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,273 @@
+
+- `Device.deviceType: string`
+
+  Returns the device's type as a string, which will be one of:
```
@ide 
https://github.com/expo/expo/pull/4682/files/e5149609e16ee6e0127ce34cb685a38d9cd89f3b#r296694851

Previously, I used an enum system but @sjchmiela suggested that I remove it. Should I bring back the enum?

vivianzzhu91 committed:
[js] remove getFreeDiskAsync and getTotalCapacity on expo-device

vivianzzhu91 committed:
[js/android] move getFreeDisk and getTotalCapacity from expo-device to expo-file-system

ninjaachibi committed:
[ios] remove TV check for carrier

Removing for now (we'll need to add something similar back when we support Catalyst for Mac)

ninjaachibi committed:
[ios] Fix ip address

ninjaachibi committed:
[ios] rename isPinOrFingerprint -> hasLocalAuthentication

per @ide's PR comment, we changed the name of this method

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,273 @@
+});
+```
+
+### `Device.getTotalDiskCapacityAsync()`
````
How about the `totalMemory` in the constants? Shall we move that as well?

ninjaachibi committed:
[ios] move storage methods to expo-file-system

@ide suggested that we move storage related methods ot the expo-file-system unimodule

vivianzzhu91 committed:
[android] fix available storage size on getTotalDiskCapacity

vivianzzhu91 committed:
[js] add stoarge functions to FileSystemModule

vivianzzhu91 committed:
[docs] add comprehensive description to the module

vivianzzhu91 committed:
[docs] add storage function descriptions

vivianzzhu91 committed:
[docs] remove storage functions

vivianzzhu91 committed:
[docs] update device docs with new functions

vivianzzhu91 committed:
[js] add new functions and constants

vivianzzhu91 committed:
[android] finish additional functions

vivianzzhu91 committed:
[docs] add example to constants

vivianzzhu91 committed:
[constants] remove deviceId and deviceName from constant module

vivianzzhu91 committed:
[js] fix bugs and update docs

ninjaachibi committed:
Update ExpoKit.podspec

add warning back

ninjaachibi committed:
Update device.md

add ANDRIOD.BUILD.ID link

vivianzzhu91 committed:
[android] add isRooted to device module

ninjaachibi committed:
[ios] set up new constants and methods

setting up new constants and methods that would be useful for the module

vivianzzhu91 committed:
[android] rename isRootedAsync to isRootedExperimentalAsync

ninjaachibi committed:
[ios] Implement new constants

Implemented the osBuildId and osVersion methods

ninjaachibi committed:
[js] Fix modelName

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+  - `"google/sdk_gphone_x86/generic_x86:9/PSR1.180720.075/5124027:user/release-keys"`
+
+- `Device.osBuildId: string`
```
Let's differentiate this more from the other build ID:
```suggestion
- `Device.osInternalBuildId: string`
```

Also, move this below `osBuildId` so they are grouped.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+- `Device.productName: string` (Android only)
+
+  Gets the device's overall product name. Chosen by the device implementer containing the development name or code name of the device.
```
```suggestion
  The device's overall product name chosen by the device implementer containing the development name or code name of the device.
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+- `Device.osBuildId: string`
+
+  A string that identifies the build number of the operating system.
```
```suggestion
  The internal build ID of the OS running on the device. On Android, this corresponds to `Build.ID`. On iOS, this is the same value as [`Device.osBuildId`](#deviceosbuildid).
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+  A string that identifies the build number of the operating system.
+
+  - `"MMB29K"`
```
```suggestion
  - Android: e.g.,`"MMB29K"`
  - iOS: e.g., `"16F203"`
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+## Methods
+
+### `Device.getPlatformFeaturesAsync()` (Android Only)
```
```suggestion
### `Device.getPlatformFeaturesAsync()` (Android only)
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,244 @@
+});
+```
+
+### `Device.hasPlatformFeatureAsync(feature)` (Android Only)
````
```suggestion
### `Device.hasPlatformFeatureAsync(feature)` (Android only)
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Returns
+
+Returns a `Promise<boolean>` that resolves the `boolean` value for whether the device has the system feature passed to the function.
```
```suggestion
Returns a `Promise<boolean>` that resolves to a `boolean` value for whether the device has the specified system feature.
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,244 @@
+await Device.hasPlatformFeatureAsync('amazon.hardware.fire_tv'); // true or false
+```
+
+### `Device.getMaxMemoryAsync()` (Android Only)
````
```suggestion
### `Device.getMaxMemoryAsync()` (Android only)
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Returns
+
+Returns a Promise that resolves the maximum available memory that the Java vm will use, in bytes.
```
```suggestion
Returns a Promise that resolves the maximum available memory that the Java VM will use, in bytes.
```

What does this method do on web and on iOS?

ninjaachibi replied:
It's listed as an Android only method at the moment.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+### `Device.getMaxMemoryAsync()` (Android Only)
+
+Returns the maximum amount of memory that the Java virtual machine will attempt to use. If there is no inherent limit then the value [Long.MAX_VALUE](https://developer.android.com/reference/java/lang/Long.html#MAX_VALUE) will be returned.
```
```suggestion
Returns the maximum amount of memory that the Java virtual machine will attempt to use. If there is no inherent limit then `Number.MAX_SAFE_INTEGER` is returned.
```

JS numbers can't fit Long.MAX_VALUE. Additionally, when possible, we want to write the documentation in terms of JS rather than Java. Make sure to change the implementation so that we get this behavior.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,244 @@
+});
+```
+
+### `Device.isSideLoadingEnabled()` (Android Only)
````
```suggestion
### `Device.isSideLoadingEnabled()` (Android only)
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+### `Device.isSideLoadingEnabled()` (Android Only)
+
+Whether applications can be installed for this user via the system's [Intent#ACTION_INSTALL_PACKAGE](https://developer.android.com/reference/android/content/Intent.html#ACTION_INSTALL_PACKAGE) mechanism.
```
```suggestion
Returns whether applications can be installed for this user via the system's [Intent#ACTION_INSTALL_PACKAGE](https://developer.android.com/reference/android/content/Intent.html#ACTION_INSTALL_PACKAGE) mechanism rather than through the OS's default app store, like Google Play.
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Returns
+
+Returns a promise that resolves to a number that represents the milliseconds since last reboot, not counting time spent in deep sleep.
```
Is the note about deep sleep true for all platforms (Android, iOS, web) or just some of them?

vivianzzhu91 replied:
iOS also only counts only the time being turned on.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+### `Device.isRootedExperimentalAsync()` (Android only)
+
+Checks whether the device has been rooted. It is not surely reliable because it's hard to test all rooted devices thoroughly.
```
```suggestion
Checks whether the device has been rooted. It is not completely reliable because it's hard to test all rooted devices thoroughly.
```

> because it's hard to test all rooted devices thoroughly.

I don't understand what this means and what the specific challenges are. Think about this from the perspective of a finance app trying to prevent fraud and explain the threat model to them.

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Returns
+
+Returns a promise that resolves to a boolean that whether this device is rooted.
```
```suggestion
Returns a promise that resolves to a boolean that specifies whether this device is rooted. On iOS and web, the boolean is always `false`.
```

(or specify some other behavior for non-Android platforms)

ninjaachibi committed:
Add getUptimeAsync

add getUptimeAsync function

ide commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -323,6 +323,40 @@ Returns a Promise that resolves to an object with the following fields:
 
+### `FileSystem.getFreeDiskStorageAsync()`
+
+Gets available internal disk storage size, in bytes. This will get you the free space on the data partition. The data partition hosts all the internal storage for all apps on the device.
```
```suggestion
Gets the available internal disk storage size, in bytes. This returns the free space on the data partition that hosts all of the internal storage for all apps on the device.
```

ide commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -323,6 +323,40 @@ Returns a Promise that resolves to an object with the following fields:
+
+#### Returns
+
+Returns a Promise that resolves to a number that represents the available internal disk storage size in bytes.
```
```suggestion
Returns a Promise that resolves to the number of bytes available on the internal disk.
```

ide commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -323,6 +323,40 @@ Returns a Promise that resolves to an object with the following fields:
+
+#### Returns
+
+Returns a Promise that resolves to a number that represents the total internal disk storage capacity in bytes.
```
```suggestion
Returns a Promise that resolves to a number that specifies the total internal disk storage capacity in bytes.
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+#### Returns
+
+A Promise that resolves to an array of strings, each containing the name of the available feature on the current device system.
```
```suggestion
A Promise that resolves to an array of strings, each of which is a platform-specific name of an available feature on the current device.
```

ide commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+### `Device.getPlatformFeaturesAsync()` (Android Only)
+
+Get a list of features that are available on the system.
```
```suggestion
Get a list of features that are available on the system. The feature names are platform-specific.
```

ide reviewed (changes requested):
I reviewed just the API and docs, no implementation yet -- left comments inlined.

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+- `Device.brand: string`
+
+  Gets the device brand. The consumer-visible brand with which the product/hardware will be associated, if any.
```
It should be non-null.

ninjaachibi committed:
Add isRootedExperimentalAsync() method

This is pretty experimental, needs to be tested. Referenced from https://github.com/thii/DTTJailbreakDetection/blob/master/Classes/DTTJailbreakDetection.m

ninjaachibi commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+- `Device.totalMemory: number`
+
+  Gets the device's total memory, in bytes. It means the total memory accessible by the kernel. This is basically the RAM size of the device, not including below-kernel fixed allocations like DMA buffers, RAM for the baseband CPU, etc.
```
It returns `@(0)` if we can't grab the total memory on the objective C side. What about android @vivianzzhu91

vivianzzhu91 replied:
It is a system constant, referenced from here https://developer.android.com/reference/android/app/ActivityManager.MemoryInfo. So I just return this constant and I'm not sure if this would return null or not.

ninjaachibi commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,244 @@
+
+  `true` if the app is running on a device, `false` if running in a simulator or emulator.
+
+- `Device.deviceType: string`
```
Didn't know that was a thing. Fascinating

vivianzzhu91 committed:
[docs] update device docs description

ninjaachibi committed:
[ios] fix isRooted, systemUptime

vivianzzhu91 committed:
[docs] update device and filesystem docs

vivianzzhu91 committed:
[js] update constants with device.md

ninjaachibi committed:
[ios] Update methods

* add osInternalBuildId
* rename supportedCPUArchitectures -> supportedCpuArchitectures
* refactor deviceType as a method

vivianzzhu91 committed:
[android] update related functions and constants

vivianzzhu91 committed:
[js] make platform specific variables to be nullable

ninjaachibi committed:
[iOS] capitalize deviceType enum values

ninjaachibi committed:
[ios] update pods

ninjaachibi committed:
[docs] update description for isRootedExperimentalAsync()

ninjaachibi committed:
[js] set up enums for DeviceType

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,116 @@
+
+import { deviceNamesByCode } from './DeviceNameByCode';
+import { Platform, UnavailabilityError } from '@unimodules/core';
+import { DeviceType } from './Device.types';
```
```suggestion
import { DeviceType } from './Device.types';

```

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,116 @@
+import ExpoDevice from './ExpoDevice';
+
+import { deviceNamesByCode } from './DeviceNameByCode';
+import { Platform, UnavailabilityError } from '@unimodules/core';
```
Move to the top

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,116 @@
+    case deviceType === DeviceType.TABLET: return DeviceType.TABLET;
+    case deviceType === DeviceType.TV: return DeviceType.TV;
+    case deviceType === DeviceType.DESKTOP: return DeviceType.DESKTOP;
+    default: return DeviceType.UNKNOWN;
```
Is this `switch` necessary? Maybe you could use something like `if (!Object.values(DeviceType).includes(deviceType)) { return DeviceType.UNKNOWN; }`?

vivianzzhu91 replied:
Because we returned string from the native codes (iOS and Android), so we might want to check which enum type the returned string belongs to.

sjchmiela commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,116 @@
+import { DeviceType } from './Device.types';
+export const brand = ExpoDevice ? ExpoDevice.brand : null;
+export const manufacturer = ExpoDevice ? ExpoDevice.manufacturer : null;
+let model;
```
Couldn't we `export let model` and change its value later? 🤔

sjchmiela commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -520,6 +523,35 @@ public void onResponse(Call call, Response response) throws IOException {
+      long blockCount = root.getBlockCountLong();
+      long blockSize = root.getBlockSizeLong();
+      BigInteger capacity = BigInteger.valueOf(blockCount).multiply(BigInteger.valueOf(blockSize));
+      promise.resolve(capacity.doubleValue());
```
How big could BigInteger be? Can it overflow?

vivianzzhu91 replied:
I cast it down to `Number.MAX_SAFE_INTEGER` if it exceeds the max integer JS could get.

sjchmiela commented on packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m:
```diff
@@ -757,6 +767,11 @@ - (void)sendEventWithName:(NSString *)eventName body:(id)body
 }
 
+- (NSDictionary *)documentFileSystemAttributes {
+  //  NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
```
```suggestion
```

sjchmiela commented on yarn.lock:
```diff
@@ -19,29 +19,29 @@
-  version "7.2.2"
-  resolved "https://registry.yarnpkg.com/@babel/core/-/core-7.2.2.tgz#07adba6dde27bb5ad8d8672f15fde3e08184a687"
-  integrity sha512-59vB0RWt09cAct5EIe58+NzGP4TFSD3Bz//2/ELy3ZeTeKF6VTD1AXlH8BGGbCX0PuobZBsIzO7IAI9PH67eKw==
+"@babel/core@7.4.3":
```
Interesting that `yarn.lock` changed… Does running `yarn` on `master` also render such results?

sjchmiela commented on packages/expo-file-system/src/FileSystem.ts:
```diff
@@ -135,6 +135,19 @@ export async function readDirectoryAsync(fileUri: string): Promise<string[]> {
+    throw new UnavailabilityError('expo-file-system', 'getFreeDiskStorageAsync');
+  }
+  return await ExponentFileSystem.getFreeDiskStorageAsync();
+}
```
```suggestion
}

```

sjchmiela reviewed (commented):
Another review round! ⭐️

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,263 @@
+
+* `Device.supportedCpuArchitectures: string[]`
+
+  Returns a list of supported processor architecture versions. The device expects the binaries it runs to be compiled for one of these architectures. The returned list will be empty if the supported architectures could not be determined.
```
You are right, it should be null if there's none, although it's not specified in the official docs.

ninjaachibi committed:
[ios] make CpuType camelcase

vivianzzhu91 committed:
[docs/js] update code and docs format

vivianzzhu91 committed:
[android] make supportedabis to be nullable

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,263 @@
+**Examples**
+
+```js
+await Device.getPlatformFeaturesAsync().then(allFeatures => {
````
We decide to go for the preferred `await`!

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -323,6 +323,40 @@ Returns a Promise that resolves to an object with the following fields:
 
 - **uri (_string_)** -- A `content://` URI pointing to the file. This is the same as the `fileUri` input parameter but in different format.
 
+### `FileSystem.getFreeDiskStorageAsync()`
```
Maybe follow other expo header templates for now since the return types are clearly explained in the `Returns` field.

ninjaachibi committed:
Merge branch 'master' into @expo-device

ninjaachibi committed:
[ios] update pods

vivianzzhu91 committed:
[android] convert limit of Big Integer in storage related functions

vivianzzhu91 committed:
[js] update DeviceType enum syntax and move deviceYearClass from constant module

ninjaachibi committed:
[ios] Add deviceYearClass constant

vivianzzhu91 committed:
move deviceYearClass from expo-constant to expo-device

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/constants.md:
```diff
@@ -26,10 +26,6 @@ The version string of the Expo client currently running.
-
-A human-readable name for the device type.
-
 ### `Constants.deviceYearClass`
```
Done!

ninjaachibi committed:
[ios] remove deviceYear from expo-constants

vivianzzhu91 committed:
[tests] add tests on iOS and Android platforms for device module

vivianzzhu91 committed:
[tests] add test suites for device module

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,316 @@
+
+- (NSString *)osBuildId {
+#if TARGET_OS_TV
+  return @"not available";
```
yep

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDeviceUID.m:
```diff
@@ -0,0 +1,143 @@
+//  Created by Robert Luo on 6/18/19.
```
I think we will remove the remove `EXDeviceUID` and `uniqueId` altogether and move it to a new `expo-application` module that exposes constants regarding the application itself.

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,316 @@
+}
+
+@property (nonatomic, strong) NSString *webViewUserAgent;
+@property (nonatomic) bool isEmulator;
```
removed the `isEmulator` property in favor of the existing `isDevice` property from `expo-constants`

ninjaachibi committed:
[ios] update ios methods

PR feedback

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,316 @@
+  __weak EXDevice *weakSelf = self;
+  EXDevice *strongSelf = weakSelf;
+  
+  __block WKWebView *webView;
```
@sjchmiela but isn't `strongSelf` an undeclared identifier in the condition of the the `if` since it isn't defined until we're inside the `if` statement?

ninjaachibi replied:
unless if change ` if (!strongSelf.webViewUserAgent)` to ` if (_webViewUserAgent)`

vivianzzhu91 committed:
[test] move testsuite to new branch

ninjaachibi committed:
[ios] update getUserAgentAsync

update strongself and weakself references

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,316 @@
+  // retrieve the current interfaces - On success, returns 0; on error, -1 is returned, and errno is set appropriately.
+  success = getifaddrs(&interfaces);
+  
+  if (success == 0) {
```
yep that makes sense!

ninjaachibi committed:
[ios] update getIpAddressAsync

since 0 means success, the variable should be named error. 0 meaning that there's no error

ninjaachibi committed:
Update package.json

Add newline character

esamelson commented on docs/pages/versions/unversioned/sdk/filesystem.md:
```diff
@@ -323,6 +323,40 @@ Returns a Promise that resolves to an object with the following fields:
 
 - **uri (_string_)** -- A `content://` URI pointing to the file. This is the same as the `fileUri` input parameter but in different format.
 
+### `FileSystem.getFreeDiskStorageAsync()`
```
@ninjaachibi @vivianzzhu91 it looks like the error handling behavior in both of these functions differs across platforms -- on Android if an exception is caught, the promise is rejected, whereas on iOS the promise is never rejected but might resolve with nil (undefined). Could you standardize this behavior across platforms and document whichever option you choose? Which one you choose should probably depend on whether you expect the exception/null case to happen in any reasonable scenarios -- if so it might good for the user to expect to handle an occasional `null`/`undefined` value. Otherwise, rejecting the promise would signify that something extra bizarre has happened.

esamelson reviewed (changes requested):
Thanks for continuing to iterate on this, both of you! The implementation is looking good, I added just a few comments there, and a few more comments on the docs.

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,262 @@
+
+- `Device.brand: string`
+
+  Gets the device brand. The consumer-visible brand of the product/hardware.
```
Sure! Maybe we should stick to the constant value itself and remove all verbs.

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,262 @@
+
+- `Device.productName: string` (Android only)
+
+  The device's overall product name chosen by the device implementer containing the development name or code name of the device. Corresponds to [`Build.PRODUCT`](https://developer.android.com/reference/android/os/Build#PRODUCT).
```
This is already linked with android docs about `Build.PRODUCT`.

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,262 @@
+  // false/true
+```
+
+### `Device.getDeviceTypeAsync()`
````
James mentioned about the phone might be a tablet now, so it might change? https://github.com/expo/expo/pull/4682#discussion_r302638991

vivianzzhu91 commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,130 @@
+import ExpoDevice from './ExpoDevice';
+
+import { deviceNamesByCode } from './DeviceNameByCode';
+import { DeviceType as _DeviceType } from './Device.types';
```
Because in typescript exporting an imported enum types would have different syntax, so if exporting them with the same name, it wouldn't work. Referenced from https://stackoverflow.com/questions/47701270/re-export-typescript-enum-from-namespace.

vivianzzhu91 commented on packages/expo-device/src/Device.ts:
```diff
@@ -0,0 +1,130 @@
+  const deviceType = await ExpoDevice.getDeviceTypeAsync();
+  switch (deviceType) {
+    case _DeviceType.PHONE:
+      return Promise.resolve(_DeviceType.PHONE);
```
Got it!

vivianzzhu91 committed:
update docs format and java exception handler

vivianzzhu91 committed:
update docs and native code format

vivianzzhu91 committed:
[docs] unify example format

vivianzzhu91 committed:
[web] add some web constant functions

ninjaachibi committed:
[js/ts] fix modelId

fixed typo in the ts file that made modelId undefined in thests

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDevice.m:
```diff
@@ -0,0 +1,460 @@
+           @"modelId": UMNullIfNil([self deviceId]),
+           @"isDevice": @([self isDevice]),
+           @"manufacturer": @"Apple",
+           @"supportedCpuArchitectures": @[[self getCpuType]],
```
it shouldn't fail, but i can return `NULL`, I've updated the code to reflect this.

ninjaachibi committed:
[ios] update EXDevice.m

* add `null` return check for getCpuType
* remove EXDeviceUID class

ninjaachibi committed:
[ios] remove methods

removed carrier, getUserAgent, hasLocalAuthentication, getMacAddress, getIpAddress since we're moving these to different modules

vivianzzhu91 committed:
[filesystem] fix filesystem broken codes

ninjaachibi committed:
[docs] Update docs

ninjaachibi committed:
[ios] revert changes to xcode config files

ninjaachibi committed:
Update project.pbxproj

vivianzzhu91 commented on packages/expo-device/src/ExpoDevice.web.ts:
```diff
@@ -0,0 +1,5 @@
+export default {
+  get name(): string {
```
Added some functions that might be available on the web.

ninjaachibi committed:
Update project.pbxproj

ran pod install

ninjaachibi committed:
[ios] Add error handling in FileSystem

Added promise rejections to be consistent with Android code

vivianzzhu91 committed:
[web] add web function return types with undefined

vivianzzhu91 committed:
reset yarn.lock to resolve merge conflict

ninjaachibi committed:
Update device.md

ninjaachibi committed:
Revert "[ios] remove deviceYear from expo-constants"

This reverts commit b7cce3fc508567fcb54d377f7640485a4f00c68b.

vivianzzhu91 committed:
[constants] revert all changes to constant module

ninjaachibi committed:
[ios] Fix error handling on filesystem

NSException likely won't be thrown, so let's reject the promise if the method returns nil

ninjaachibi committed:
Merge branch 'master' into @expo-device

esamelson reviewed (changes requested):
this is looking great @vivianzzhu91 and @ninjaachibi . Just two more small comments and then we can go ahead and land 👍

vivianzzhu91 committed:
[ios] remove uid files

vivianzzhu91 committed:
[js] move DeviceType enum from types to Device.ts

ninjaachibi commented on packages/expo-device/ios/EXDevice/EXDeviceUID.h:
```diff
@@ -0,0 +1,7 @@
+//  Created by Robert Luo on 6/18/19.
```
yep

ninjaachibi committed:
Merge branch 'master' into @expo-device

esamelson reviewed (approved):
looks good to me, thanks both of you! I think @ide wanted to take one more pass before you merge though

ninjaachibi committed:
Update device.md

vivianzzhu91 commented on docs/pages/versions/unversioned/sdk/device.md:
```diff
@@ -0,0 +1,352 @@
+
+### `Device.deviceYearClass`
+
+The [device year class](https://github.com/facebook/device-year-class) of this device.
```
It shall be `null` since there's not a reliable way to get the `deviceYearClass` on web. Will add the docs.

vivianzzhu91 committed:
[docs] add other platform behaviors and cleanup format

vivianzzhu91 committed:
[web] add all nullable constants

esamelson commented on docs/pages/versions/unversioned/sdk/device.md:
````diff
@@ -0,0 +1,352 @@
+Device.designName; // iOS: null; Android: "kminilte"
+```
+
+### `Device.productName`
````
if we're sure it really does serve the same purpose as `modelId` (a unique identifier for the device's hardware model) then I think folding it into `modelId` makes sense. I wasn't able to find a concrete explanation of what this field means to Android, though, and unless we're really sure, I'm inclined to think that it's better to keep them separate since they will both be such platform-specific values.

vivianzzhu91 committed:
[android] delete unused permission request

ide commented on packages/expo-device/src/ExpoDevice.web.ts:
```diff
@@ -0,0 +1,62 @@
+  get osBuildFingerprint(): null {
+    return null;
+  },
+  get designName(): null {
```
For fields that only Android sends back (and the intent is they are Android-only fields, rather than not-iOS fields), remove them from the web implementation of the platform layer and have the API layer handle these missing fields (ex: `ExpoDevice ? ExpoDevice.exampleField || null : null`)

ide commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -520,6 +523,45 @@ public void onResponse(Call call, Response response) throws IOException {
+      long blockCount = root.getBlockCountLong();
+      long blockSize = root.getBlockSizeLong();
+      BigInteger capacity = BigInteger.valueOf(blockCount).multiply(BigInteger.valueOf(blockSize));
+      Double capacityDouble = capacity.doubleValue();
```
```suggestion
      Double capacityDouble = Math.max(capacity.doubleValue(), Math.pow(2, 53) - 1);
```

ide commented on packages/expo-file-system/android/src/main/java/expo/modules/filesystem/FileSystemModule.java:
```diff
@@ -520,6 +523,45 @@ public void onResponse(Call call, Response response) throws IOException {
+      BigInteger storage = BigInteger.valueOf(availableBlocks).multiply(BigInteger.valueOf(blockSize));
+      Double storageDouble = storage.doubleValue();
+      //cast down to avoid overflow
+      if (storage.longValue() > Math.pow(2, 53) - 1) {
```
math.max here too

vivianzzhu91 committed:
[docs] cleanup format and description

vivianzzhu91 committed:
update error codes and add exception handler to android

ninjaachibi committed:
[iOS] style cleaning

* change getCpuType -> cpuType
* add nullable annotation to cputype
* style fix for isRootedExperimentalAsync()

vivianzzhu91 committed:
[js] remove platform specific codes from web and add double ternary

ninjaachibi committed:
Merge branch 'master' into @expo-device

ninjaachibi committed:
[ios] handle null case in cputype

info was NULL an error would be thrown

ide committed:
Update formatting and error handling for isRootedExperimentalAsync on Android

ide committed:
Fix formatting in iOS EXDevice

ide committed:
[device] Fix bug in web implementation of platform layer

ide reviewed (approved):
This looks almost ready for beta.

vivianzzhu91 committed:
rebuild ExpoDevice.web.js and update js files

ide commented:
Make sure to squash the commits and amend the commit message (ex: it still references "phone carrier") before landing.
